enhancements/ingress/set-delete-http-headers.md

```diff
@@ -0,0 +1,1031 @@
+
+### Drawbacks
+
+## Design Details
+
+### Open Questions
+
+#### Will this enhancement cover setting or deleting request headers in addition to response headers?
+  Answer: Yes, we will be covering setting or deleting of request headers.
+  This is required for the following use-case.  
+  
+  As a cluster admin or route owner, I want to enable mutual TLS but replace the "x-ssl-client-der" request header
+  that OpenShift router adds with an "x-forwarded-client-cert" request header that my application checks for, 
+  so that I can migrate my applications that rely on mutual TLS and use the latter header name. 
+
+  The following example replaces "x-ssl-client-der" with "x-forwarded-client-cert":
+```yaml
+spec:
+  httpHeaders:
+    actions:
+      response:
+      - name: X-SSL-Client-Cert
+        action:
+        type: Set
+        set:
+          value: "%{+Q}[ssl_c_der,base64]"
+      - name:   Content-Language
+        action:
+        type: Delete
+```
+
+#### Will this enhancement cover setting the load-balancing algorithm based on headers?
+  Answer:The HAProxy configuration provides a [load balancing algorithm](http://cbonte.github.io/haproxy-dconv/2.6/configuration.html#4.2-balance) which relies on the
+  check against the incoming request's header values.
+  
+  Even though this enhancement adds the header manipulation capability it
+  doesn't aim at providing (or enhancing) the load balancing API for the IngressController or Route. 
+  
+  Therefore, such a load balancing algorithm is out of scope for this enhancement and should be handled with a separate RFE and thereby it will be considered as a non-goal.
+
+#### Will this enhancement cover deleting or appending headers?
+  Answer: Yes and no.  This enhancement will cover deletion but not appending of
+  HTTP response headers as per the requirements from
+  [RFE-464](https://issues.redhat.com/browse/RFE-464). 
+  While this RFE mentions use-cases that require the ability to delete headers, there is no use-case
+  from this RFE that requires the ability to append headers. 
+
+  Once such a use-case arises, we may add the option to append headers, with due
+  consideration of any additional requirements deriving from the use-case
+  itself or considerations around the handling of comma-separated values.  (Note
+  that a use-case involving the `x-forwarded-*` headers, which is not mentioned
+  on the RFE, is described below.)
+  
+  "set-header" does the same as "add-header" except that the header name is first removed if it existed. This is 
+  useful when passing security information to the server, where the header must not be manipulated by external users.
+  So, that is the reason we don't need an add-header.
+  References - https://wohugb.gitbooks.io/haproxy/content/config/http-response.html
+       https://www.haproxy.com/documentation/hapee/latest/traffic-routing/rewrites/rewrite-responses/
```

Comment:
```suggestion
  "set-header" can be used for the same purpose as "append-header" except that the user must know the value before adding to it. This is 
  useful when passing security information to the server, where the header must not be manipulated by external users.  We chose to use "set-header" instead of "append-header" to provide a measure of security.
  References - https://wohugb.gitbooks.io/haproxy/content/config/http-response.html
       https://www.haproxy.com/documentation/hapee/latest/traffic-routing/rewrites/rewrite-responses/
```

Reply:
The original statement is more accurate: `set-header` tells HAProxy to overwrite any existing header; that is, remove the header if it is already set and then add the header with the custom value.  The user does not need to know the old value to overwrite it with the custom value.  

By the way, is the comment about security important?  It seems confusing to me; if the point about security is really important, it should be clarified, and otherwise it can be omitted.

Reply:
ok removed it. that sounds right.

Reply:
We already use add-header in https://github.com/openshift/router/blob/master/images/router/haproxy/conf/haproxy-config.template#L554 and lines after, dealing with x-forward and forward values.  This section doesn't make sense. The question should just be  `Will this enhancement cover appending headers` and no mention of deleting, which is already covered.